src/Compilation.zig

```diff
@@ -1028,7 +1029,7 @@ pub fn create(gpa: Allocator, options: InitOptions) !*Compilation {
 
         const include_compiler_rt = options.want_compiler_rt orelse needs_c_symbols;
 
-        const must_single_thread = target_util.isSingleThreaded(options.target);
+        const must_single_thread = options.target.isWasm() and !options.linker_shared_memory;
```

Comment:
Should single-threaded remain the default for wasm32-wasi, since threading is experimental?

Reply:
I think that's the better default for now, yes. Perhaps it should check for both `--shared-memory` *and* `-fno-single-threaded` to enable multi-threaded wasm32-wasi.

Reply:
Also I think `-fno-single-threaded` and not `--shared-memory` should be a compile error explaining the incompatibility.

Alternatively, --shared-memory should become a flag (`-fshared-memory`, `-fno-shared-memory`) with a default that is affected by whether single-threaded is enabled.

Reply:
Good idea, I'll add the compatibility check with a compile error. I think option A makes the most sense to a WebAssembly user. I cannot think of a use case where you deliberately turn off `shared-memory` as that's the default. Turning it off in a case where it is enabled by default (e.g. `-fno-single-threaded`) would result in a compile-error.